### PR TITLE
[feature](function)Support negative index for function split_part

### LINKED ISSUE #13712: [Feature] split_part函数支持用负数倒取字段
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

VARCHAR split party (VARCHAR content, VARCHAR delimiter, INT field)
当前第三个参数 不支持使用负数倒取第几个字段，需要支持

https://doris.apache.org/docs/dev/sql-manual/sql-functions/string-functions/split_part?_highlight=split_part

### Use case

期望
select split_part("hello world", " ", -1);
输出： world

### Related issues

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Yulei-Yang**
# Proposed changes

Issue Number: close #13712

Returns the specified partition by splitting the string according to the delimiter. If field is positive, counting from the beginning of content,  otherwise from the ending.

mysql> select split_part('prefix_string', '\_', -1);
+--------------------------------------+
| split_part('prefix_string', '\_', -1) |
+--------------------------------------+
| string                               |
+--------------------------------------+

select split_part('prefix_string', '\_', -2);
+--------------------------------------+
| split_part('prefix_string', '\_', -2) |
+--------------------------------------+
| prefix                               |
+--------------------------------------+

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.47 seconds
 load time: 571 seconds
 storage size: 17154875741 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221102162329_clickbench_pr_38768.html

**Comment by Yulei-Yang** [comment]
@LOVEGISER could you please review it? thx

**Comment by Yulei-Yang** [comment]
@Gabriel39 could you please review it? thx

File: be/src/vec/functions/function_string.h [hunk 1]
```diff
@@ -1278,63 +1278,58 @@ class FunctionSplitPart : public IFunction {
...
-                    pre_offset = offset;
+                std::vector<int> find(str.size, -1); //store delimiter position
+                while (num < str.size) {
                     size_t n = str.size - offset - delimiter.size;
-                    char* pos = reinterpret_cast<char*>(memmem(str.data + offset + delimiter.size,
-                                                               n, delimiter.data, delimiter.size));
-                    if (pos != nullptr) {
-                        offset = pos - str.data;
-                        num++;
+                    if (delimiter.size == 1) {
```
Comment (Gabriel39) [inline be/src/vec/functions/function_string.h:1287]: Keep origin logics. We should try to not make this if-else inside a loop

**Review by Gabriel39 — COMMENTED** [review]
Why `split_part('prefix_string', '', -1)` returns `string`? I think is should be `split_part('prefix_string', '_', -1)`

**Comment by Yulei-Yang** [comment]
> oh, it's a github editor issue, which treat the content as markdown. so the underscore character is used as italic mark. I escaped this char, and it looks right now.

File: be/src/vec/functions/function_string.h (hunk 1 above)
Comment (Yulei-Yang) [reply be/src/vec/functions/function_string.h:1287]: Thank you very much for your review. While I have a question here. If I keep origin logic, there will be two parts of code which looks similar and seems duplicated, like line 1282-1284 and 1317-1333。Can you please give me an advice which issue is more critical?

Comment (Gabriel39) [reply be/src/vec/functions/function_string.h:1287]: Maybe macro is what you need

Comment (Yulei-Yang) [reply be/src/vec/functions/function_string.h]: > Keep origin logics. We should try to not make this if-else inside a loop

I get your point. It's about efficiency, if/else in a loop will prevent loop unrolling. I update the code according to your opinion. Please review, thx in advance.

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Pull request merged**